crates/sqlbuiltins/src/functions.rs

```diff
@@ -104,6 +104,53 @@ pub trait TableFuncContextProvider: Sync + Send {
     fn get_credentials_entry(&self, name: &str) -> Option<&CredentialsEntry>;
 }
 
+/// Value from a function parameter.
+#[derive(Debug, Clone)]
+pub enum FuncParamValue {
```

Comment:
IMO, this is kinda hard to understand the flow. Couldn't we just pass around the entire `Expr` and do a `string_from_expr` instead of `string_from_scalar`? 


over in polars, I created a trait to extract values from `Expr`

```rust

pub(crate) trait FromSqlExpr {
    fn from_sql_expr(expr: &SqlExpr, ctx: &SQLContext) -> PolarsResult<Self>
    where
        Self: Sized;
}

impl FromSqlExpr for String {
    fn from_sql_expr(expr: &SqlExpr, _: &SQLContext) -> PolarsResult<Self>
    where
        Self: Sized,
    {
        match expr {
            SqlExpr::Value(v) => match v {
                SqlValue::SingleQuotedString(s) => Ok(s.clone()),
                _ => polars_bail!(ComputeError: "can't parse literal {:?}", v),
            },
            _ => polars_bail!(ComputeError: "can't parse literal {:?}", expr),
        }
    }
}
```

could likely do something similar here. so instead of 
```rust
string_from_scalar(args.next().unwrap().try_into()?)?;
```
we'd end up with
```rust
String::try_from_sql_expr(args.next.unwrap())?
```
which I feel is much more readable.

Reply:
That's a neat idea. I did it to explicitly differentiate between `Ident` and `ScalarValue`, so a little confused how should I implement the trait. Maybe have some default functions...

```rust
pub trait FromFuncParamValue {
    fn try_from_scalar(s: ScalarValue) -> Result<Self>;
    fn try_from_ident(s: String) -> Result<Self>;

    fn from_scalar(v: FuncParamValue) -> Result<Self> {
        match v {
            FuncParamValue::Scalar(s) => Self::try_from_scalar(s),
            _ => Err(...),
        }
    }

    fn from_ident(v: FuncParamValue) -> Result<Self> {
         match v {
            FuncParamValue::Ident(s) => Self::try_from_ident(s),
            _ => Err(...),
        }
    }
}
```

So we'll be able to:

```rust
String::from_scalar(args.next().unwrap())?
```

Reply:
I think the idea is to defer the parsing of `Expr -> ScalarValue` and `Expr -> Ident`. We can just pass along the `Expr` 
instead of `args: Vec<FuncParamValue>,`
`args: Vec<Expr>,`

then just extract the needed value there. 

for example

```rust
async fn create_provider(
        &self,
        _: &dyn TableFuncContextProvider,
        args: Vec<FuncParamValue>,
        _opts: HashMap<String, FuncParamValue>,
    ) -> Result<Arc<dyn TableProvider>> {
        match args.len() {
            3 => {
                let mut args = args.into_iter();
                let conn_str = string_from_scalar(args.next().unwrap().try_into()?)?;
                let schema = string_from_scalar(args.next().unwrap().try_into()?)?;
                let table = string_from_scalar(args.next().unwrap().try_into()?)?;
               //...
            }
            _ => Err(BuiltinError::InvalidNumArgs),
        }
    }
```

would be something like this

```rust
async fn create_provider(
        &self,
        _: &dyn TableFuncContextProvider,
        args: Vec<Expr>,
        _opts: HashMap<String, Expr>,
    ) -> Result<Arc<dyn TableProvider>> {
        match args.len() {
            3 => {
                let mut args = args.into_iter();
                let conn_str = String::try_from_sql_expr(args.next().unwrap())?
                let schema = String::try_from_sql_expr(args.next().unwrap())?
                let table = String::try_from_sql_expr(args.next().unwrap())?
               // ...
            }
            _ => Err(BuiltinError::InvalidNumArgs),
        }
    }
```
and the `FromSqlExpr` would match on both scalar strings & ident
    
```rust
impl FromSqlExpr for String {
    fn from_sql_expr(expr: &SqlExpr, _: &SQLContext) -> PolarsResult<Self>
    where
        Self: Sized,
    {
        match expr {
           SqlExpr::Identifier(ident) => Ok(self.normalizer.normalize(ident).into())
            SqlExpr::Value(v) => match v {
                SqlValue::SingleQuotedString(s) => Ok(s.clone()),
                _ => polars_bail!(ComputeError: "can't parse literal {:?}", v),
            },
            _ => polars_bail!(ComputeError: "can't parse literal {:?}", expr),
        }
    }
}
```

Reply:
The reason I'm not willing to do this is because it doesn't let us differentiate between the source of parameter.

For example:

```rust
let credentials: String = args.next().unwrap().from_sql_expr()?;
```

This will be parsed for both the cases, when the `Expr` is `Ident` or a `ScalarValue`. Basically, doesn't let us differentiate between a literal and an identifier, so both the following would work:

```
select * from csv_scan('gs://bucket/object', mycreds);

select * from csv_scan('gs://bucket/object', 'mycreds');
```

I understand that we don't defer parsing `Expr` in this case. What I think of this is, we eventually convert an `Expr` to a `FuncParamValue` (equivalent to an `Expr`).

If the above case I mentioned doesn't seem like a big deal, I think your method is better, but I am only confused about that. Will give it some thought today, would love to hear your opinions. Also tagging @scsmithr

Reply:
I thought about it; maybe it doesn't matter. If we ever need a strict distinction, we can create a wrapper over `String` like `Ident(String)` and implement the trait.

I will copy stuff from your PR to minimize conflict resolution...

Reply:
I think we still need something like `FuncParamValue` as in intermediate since we want named parameters as options that we can get by name. We can't force positions for named arguments. This is something we'll have to process before-hand.

Reply:
Pushed a change. I feel this is fine, having an intermediate `FuncParamValue`. It's much clearer to read with the trait `FromFuncParamValue`. I don't see any disadvantage of going this way since using this intermediate, we eventually just figure out during planning that the arguments are valid and something our functions will accept.

Reply:
To throw in my input, I would _prefer_ if we were able to differentiate between idents and strings, but I think having a clear implementation is more important here.

I didn't have a chance to go through the previous iteration, but the code we have here looks fine to me. However I want to leave end decision up to you two.

Reply:
I am quite happy with this. I guess we can merge this for now. We'll make more changes later if required.